csrc/selective_scan/selective_scan_bwd_kernel.cuh

```diff
@@ -515,14 +525,15 @@ void selective_scan_bwd_launch(SSMParamsBwd &params, cudaStream_t stream) {
     });
 }
 
+// TODO: make conditional. What's the best practice?
```

Comment:
It depends.  Some kernels are written such that they can successfully adapt to changing the warp size or multiple of warp size.  Here, as a template param, we must instantiate both 32 and 64 variants because ROCm supports both kinds of GPUs, and add a runtime check for the current device's warp size to pick the right kernel.  Since this is a pytorch extension, we can leverage what pytorch already provides.  If there is concern about upstreaming this, you could #ifdef USE_ROCM so that cuda builds don't instantiate kernels that won't be used.  See both examples below.  Not sure which option is better or less confusing.

```C++
if (at::cuda::warp_size() == 32) {
    if (params.seqlen <= 128) {
        selective_scan_bwd_launch<32, 4, input_t, weight_t>(params, stream);
    } else if (params.seqlen <= 256) {
        selective_scan_bwd_launch<32, 8, input_t, weight_t>(params, stream);
    } else if (params.seqlen <= 512) {
        selective_scan_bwd_launch<32, 16, input_t, weight_t>(params, stream);
    } else if (params.seqlen <= 1024) {
        selective_scan_bwd_launch<64, 16, input_t, weight_t>(params, stream);
    } else {
        selective_scan_bwd_launch<128, 16, input_t, weight_t>(params, stream);
    }
}
#ifdef USE_ROCM
else {
    if (params.seqlen <= 128) {
        selective_scan_bwd_launch<64, 4, input_t, weight_t>(params, stream);
    } else if (params.seqlen <= 256) {
        selective_scan_bwd_launch<64, 8, input_t, weight_t>(params, stream);
    } else if (params.seqlen <= 512) {
        selective_scan_bwd_launch<64, 16, input_t, weight_t>(params, stream);
    } else if (params.seqlen <= 1024) {
        selective_scan_bwd_launch<64, 16, input_t, weight_t>(params, stream);
    } else {
        selective_scan_bwd_launch<128, 16, input_t, weight_t>(params, stream);
    }
}
#endif
```

Other option:

```C++
    if (params.seqlen <= 128) {
        if (at::cuda::warp_size() == 32) {
            selective_scan_bwd_launch<32, 4, input_t, weight_t>(params, stream);
        }
        else {
            selective_scan_bwd_launch<64, 4, input_t, weight_t>(params, stream);
        }
    } else if (params.seqlen <= 256) {
        if (at::cuda::warp_size() == 32) {
            selective_scan_bwd_launch<32, 8, input_t, weight_t>(params, stream);
        else {
            selective_scan_bwd_launch<64, 8, input_t, weight_t>(params, stream);
        }
    } else if (params.seqlen <= 512) {
        if (at::cuda::warp_size() == 32) {
            selective_scan_bwd_launch<32, 16, input_t, weight_t>(params, stream);
        }
        else {
            selective_scan_bwd_launch<64, 16, input_t, weight_t>(params, stream);
        }
    } else if (params.seqlen <= 1024) {
        selective_scan_bwd_launch<64, 16, input_t, weight_t>(params, stream);
    } else {
        selective_scan_bwd_launch<128, 16, input_t, weight_t>(params, stream);
    }